File: ios/MeasureSDK/Events/Event.swift
//
//  Event.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 05/09/24.
//

import Foundation

/// Represents an event in Measure. This object maps closely to the event object in the Measure API.
final class Event<T: Codable>: Codable {
    /// A unique identifier for the event.
    let id: String

    /// The session id of the event. This is the session id of the session in which the event was triggered.
    let sessionId: String

    /// The timestamp of the event. The time when the event was triggered, measured in milliseconds since epoch.
    let timestamp: String

    /// The timestamp of the event in milliseconds.
    let timestampInMillis: Number?

    /// The type of the event.
    let type: EventType

    /// The data collected. This can be any object that conforms to `Codable`.
    let data: T?

    /// Attachments that can be added to the event.
    var attachments: [Attachment]?

    /// Additional key-value pairs that can be added to the event.
    var attributes: Attributes?

    /// A flag to indicate if the event is triggered by the user or the SDK.
    let userTriggered: Bool

    /// Attributes set by the user in the event. The type of values in the map is set to Any here, however, the allowed values can only be String, Int, Long, Double, Float or Boolean.
    let userDefinedAttributes: String?

    init(id: String,
         sessionId: String,
         timestamp: String,
         timestampInMillis: Number,
         type: EventType, data: T?,
         attachments: [Attachment]?,
         attributes: Attributes?,
         userTriggered: Bool,
         userDefinedAttributes: String? = nil) {
        self.id = id
        self.sessionId = sessionId
        self.timestamp = timestamp
        self.type = type
        self.data = data
        self.attachments = attachments
        self.attributes = attributes
        self.userTriggered = userTriggered
        self.timestampInMillis = timestampInMillis
        self.userDefinedAttributes = userDefinedAttributes
    }

    enum CodingKeys: String, CodingKey {
        case id
        case sessionId = "session_id"
        case timestamp
        case type
        case data
        case attachments
        case attributes
        case userTriggered = "user_triggered"
        case timestampInMillis
        case userDefinedAttributes = "user_defined_attributes"
    }

    func appendAttributes(_ attributeProcessors: [AttributeProcessor]) {
        attributeProcessors.forEach { processor in
            processor.appendAttributes(&self.attributes!)
        }
    }
}

File: android/measure/src/main/java/sh/measure/android/events/Event.kt
package sh.measure.android.events

import sh.measure.android.attributes.AttributeValue
import sh.measure.android.cel.CelFieldAccessor

/**
 * Represents an event in Measure. This object maps very closely to the event object in
 * the Measure API.
 */
internal data class Event<T>(
    /**
     * A unique identifier for the event.
     */
    val id: String,

    /**
     * The session id of the event. This is the session id of the session in which the event was
     * triggered.
     */
    val sessionId: String,

    /**
     * The timestamp of the event. The time when the event was triggered. Measured in milliseconds
     * since epoch.
     */
    val timestamp: String,

    /**
     * The type of the event. See [EventType] for the list of event types.
     */
    val type: EventType,

    /**
     * The data collected. This can be any object that is annotated with `@Serializable`.
     */
    val data: T,

    /**
     * Attachments that can be added to the event.
     */
    val attachments: MutableList<Attachment>,

    /**
     * Additional key value pairs that can be added to the event.
     */
    val attributes: MutableMap<String, Any?>,

    /**
     * A flag to indicate if the event is triggered by the user or the SDK.
     */
    val userTriggered: Boolean,

    /**
     * Attributes set by the user in the event. The type of values in the map is set to Any here,
     * however, the allowed values can only be String, Int, Long, Double, Float or Boolean.
     */
    val userDefinedAttributes: Map<String, AttributeValue>,
) : CelFieldAccessor {
    /**
     * Adds an attribute to the event.
     *
     * @param key The key of the attribute.
     * @param value The value of the attribute.
     */
    fun appendAttribute(key: String, value: Any?) {
        attributes[key] = value
    }

    fun addAttachment(attachment: Attachment) {
        attachments.add(attachment)
    }

    override fun getField(fieldName: String): Any? {
        return when (fieldName) {
            "id" -> id
            "session_id" -> sessionId
            "user_triggered" -> userTriggered
            "type" -> type.value
            "attribute" -> attributes
            "user_defined_attribute" -> userDefinedAttributes
            EventType.STRING.value,
            EventType.EXCEPTION.value,
            EventType.ANR.value,
            EventType.APP_EXIT.value,
            EventType.CLICK.value,
            EventType.LONG_CLICK.value,
            EventType.SCROLL.value,
            EventType.LIFECYCLE_ACTIVITY.value,
            EventType.LIFECYCLE_FRAGMENT.value,
            EventType.LIFECYCLE_APP.value,
            EventType.COLD_LAUNCH.value,
            EventType.WARM_LAUNCH.value,
            EventType.HOT_LAUNCH.value,
            EventType.NETWORK_CHANGE.value,
            EventType.HTTP.value,
            EventType.MEMORY_USAGE.value,
            EventType.TRIM_MEMORY.value,
            EventType.CPU_USAGE.value,
            EventType.SCREEN_VIEW.value,
            EventType.CUSTOM.value,
            EventType.BUG_REPORT.value,
                -> data
            else -> null
        }
    }
}
